src/core/matching_state.rs

```diff
@@ -1,78 +1,114 @@
-}
-
+/// State passed to heuristics to manipulate matches and query current directory contents.
+///
+/// Only this type should be used to interact with the filesystem and return meaningful heuristic result to the user.
 pub struct MatchingState<'entries> {
-    contents: HashMap<OsString, PathData<'entries>>,
+    /// Optimized storage for current direcotry contents.
+    contents: HashMap<OsString, (&'entries mut Entry, Vec<MatchDataBuilder>)>,
+    /// Path of the current directory.
+    parent_path: &'entries Path,
+    /// Current heuristic being processed.
     pub(super) current_heuristic: Option<&'static dyn Heuristic>,
-    pub important_files: &'entries mut Vec<PathBuf>,
+    /// Files inherited from parent directories, hashed by heuristic type.
+    inherited_files: &'entries mut InheritedFiles,
 }
 
 impl<'entries> MatchingState<'entries> {
-    pub(super) fn new(children: &'entries mut [&mut Entry], files: &'entries mut Vec<PathBuf>) -> Self {
+    /// Creates a new matching state for the specified directory, its entries and inherited files.
+    pub(super) fn new(
+        children: &'entries mut [&mut Entry],
+        files: &'entries mut InheritedFiles,
+        path: &'entries Path,
+    ) -> Self {
         Self {
-            contents: children
-                .iter_mut()
-                .map(|v| {
-                    (v.file_name.clone(), PathData {
-                        entry: v,
-                        weight: 0,
-                        reasons: vec![],
-                    })
-                })
-                .collect(),
+            contents: children.iter_mut().map(|v| (v.file_name.clone(), (v.deref_mut(), vec![]))).collect(),
             current_heuristic: None,
-            important_files: files,
+            inherited_files: files,
+            parent_path: path,
         }
     }
 
+    /// Function to be called after every heuristic has done its job.
+    ///
+    /// This function filters and reorganizes all collected data in order to send it to the specified channel.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the channel is closed, which should not happen in normal operation.
     pub(super) fn process_collected_data(&mut self, sender: &Sender<MatchData>) {
-        for (_, v) in self.contents.drain() {
-            if v.weight <= 0 {
+        for (_, (entry, md)) in self.contents.drain() {
+            let md: MatchDataBuilder = md.into_iter().sum();
+            if md.weight <= 0 {
                 continue;
             }
-            v.entry.read_children_path = None;
+            entry.read_children_path = None;
             let data = MatchData {
-                path: v.entry.path(),
-                weight: v.weight,
-                reasons: v.reasons,
+                path: entry.path(),
+                group: self.parent_path.to_owned(),
+                other_data: md,
             };
             sender.send(data).expect("Sender error (did UI panic?)");
```

Comment:
maybe instead of panicing just log error?

Reply:
This is the code from before you implemented decent logging - I want to remind you that this branch was made quite some time ago.

Reply:
mark it as todo for yourself :-)